### PR TITLE
[corlib] Import System.Threading.Monitor

### PR DESCRIPTION
**Pull request by ludovic-henry**
cc @akoeplinger

### CODE REVIEW

File: mono/metadata/icall-def.h [hunk 1]
```diff
@@ -873,9 +873,7 @@ ICALL(MONIT_2, "Monitor_pulse", ves_icall_System_Threading_Monitor_Monitor_pulse
 ICALL(MONIT_3, "Monitor_pulse_all", ves_icall_System_Threading_Monitor_Monitor_pulse_all)
 ICALL(MONIT_4, "Monitor_test_owner", ves_icall_System_Threading_Monitor_Monitor_test_owner)
 ICALL(MONIT_5, "Monitor_test_synchronised", ves_icall_System_Threading_Monitor_Monitor_test_synchronised)
-ICALL(MONIT_6, "Monitor_try_enter", ves_icall_System_Threading_Monitor_Monitor_try_enter)
 ICALL(MONIT_7, "Monitor_wait", ves_icall_System_Threading_Monitor_Monitor_wait)
-ICALL(MONIT_10, "enter_with_atomic_var", mono_monitor_enter_v4)
```
Comment (marek-safar) [inline mono/metadata/icall-def.h]: Do we still need register_icall_with_wrapper for it?

Comment (ludovic-henry) [inline mono/metadata/icall-def.h]: yes, as it's used in method-to-ir.c

Comment (ludovic-henry) [inline mono/metadata/icall-def.h]: Actually, we cannot remove it as it's used in `mono_marshal_get_synchronized_wrapper`

Comment (ludovic-henry) [inline mono/metadata/icall-def.h]: `enter_with_atomic_var` has been entirely removed, and it is replaced by `Monitor:Enter(obj, ref bool)`

**Comment by akoeplinger** [comment]
We're missing a bunch of argument validation (test failures). Seems referencesource/coreclr did it in native code, probably a good idea to review them to make sure we don't forget any of these.

**Comment by ludovic-henry** [comment]
@akoeplinger we are missing these argument validation because in method-to-ir.c, we optimize Monitor.Enter at https://github.com/mono/mono/pull/2933/files#diff-e7526071878d708680e4a8c5fa5a46edR6115. That has to be fixed.

**Comment by ludovic-henry** [comment]
build

**Comment by ludovic-henry** [comment]
@marek-safar could you please review and merge accordingly? Thank you!

**Comment by ludovic-henry** [comment]
build

**Pull request merged**